components/brave_rewards/browser/rewards_service_impl.cc

```diff
@@ -186,17 +189,91 @@ std::string URLMethodToRequestType(ledger::URL_METHOD method) {
+uint64_t RoundProbiToUint64(base::StringPiece probi) {
+  if (probi.size() < 18) return 0;
+  uint64_t grant = 0;
+  base::StringToUint64(probi.substr(0, probi.size() - 18), &grant);
+  return grant;
+}
+
+
+void ExtractAndLogP3AStats(const base::DictionaryValue& dict) {
+  const base::Value* probi_value =
+      dict.FindPath({"walletProperties", "probi_"});
+  if (!probi_value || !probi_value->is_string()) {
+    LOG(WARNING) << "Bad ledger state";
+    return;
+  }
+
+  // Get grants.
+  const base::Value* grants_value = dict.FindKey("grants");
+  uint64_t total_grants = 0;
+  if (grants_value) {
+    if (!grants_value->is_list()) {
+      LOG(WARNING) << "Bad grant value in ledger_state.";
+    } else {
+      const auto& grants = grants_value->GetList();
+      // Sum all grants.
+      for (const auto& grant : grants) {
+        if (!grant.is_dict()) {
+          LOG(WARNING) << "Bad grant value in ledger_state.";
+          continue;
+        }
+        const base::Value* grant_amount = grant.FindKey("probi_");
+        const base::Value* grant_currency = grant.FindKey("altcurrency");
+        if (grant_amount->is_string() && grant_currency->is_string() &&
+            grant_currency->GetString() == "BAT") {
+          // Some kludge computations because we don't want to be very precise
+          // for P3A purposes. Assuming grants can't be negative and are
+          // greater than 1 BAT.
+          const std::string& grant_str = grant_amount->GetString();
+          total_grants += RoundProbiToUint64(grant_str);
+        }
+      }
+    }
+  }
+  const uint64_t total =
+      RoundProbiToUint64(probi_value->GetString()) - total_grants;
+  RecordWalletBalanceP3A(true, total);
+}
+
+// Returns pair of string and its parsed counterpart. We parse it on the file
+// thread for the performance sake. It's should be better to remove the string
+// representation in the [far] future.
+std::pair<std::string, base::Value> LoadStateOnFileTaskRunner(
     const base::FilePath& path) {
   std::string data;
   bool success = base::ReadFileToString(path, &data);
 
   // Make sure the file isn't empty.
   if (!success || data.empty()) {
     LOG(ERROR) << "Failed to read file: " << path.MaybeAsASCII();
-    return std::string();
+    return {};
   }
-  return data;
+  std::pair<std::string, base::Value> result;
+  result.first = data;
+
+  // Save deserialized version for recording P3A and future use.
+  JSONStringValueDeserializer deserializer(data);
+  int error_code = 0;
+  std::string error_message;
+  auto value = deserializer.Deserialize(&error_code, &error_message);
+  if (!value) {
+    LOG(ERROR) << "Cannot deserialize ledger state, error code: " << error_code
+               << " message: " << error_message;
+    return result;
+  }
+
+  const auto dict = base::DictionaryValue::From(std::move(value));
+  if (!dict) {
+    LOG(ERROR) << "Corrupted ledger state.";
+    return result;
+  }
+
+  ExtractAndLogP3AStats(*dict);
```

Comment:
why do we do additional logging in `OnLedgerStateLoaded` if we have extract function here? This way we wouldn't need to change function signature

Reply:
we need an access to some class members in `OnLedgerStateLoaded`, so we cannot do that here

Reply:
I'm still not sure this is a good place to put this. For one it's used for both ledger and publisher state, but you have errors in `ExtractAndLogP3AStats` if ledger state values are missing. Secondly this file loading is going away in favor of prefs/db so seems like the callback is better as @NejcZdovc is suggesting? What do you need access to that you can get here, but not in the callback?

Reply:
and actually I'd prefer not to do any json deserialization here because technically RewardsServiceImpl doesn't know or care how the data is stored or what it contains. I think what really needs to happen here is that the metrics need to go into ledger and not rewards service because these are internal details of ledger that shouldn't be exposed to rewards service

Reply:
@NejcZdovc maybe we should create some kind of metrics LedgerClient method for this kind of stuff?

Reply:
maybe `LedgerClient::RecordMetric(const std::string& metric_name, T value)`? We don't want to make p3a a dep of ledger so I think a generic metric interface is the way to go